### PR TITLE
#2319 Refactored DynStencil into a dataclass LFRicArgStencil

### PR DESCRIPTION
**Pull request by oakleybrunt**
PR for #2319 
Part of #2235 

Decided to refactor into a dataclass (part of the dataclasses library) for clarity.

Calls to LFRicArgStencil may not list all parameters which causes an error. I think the defualt value setting in a `namedtuple` is less readable than a `dataclass`. 

A dataclass effectively outsources some methods like `__init__` and `__repr__` to the `@dataclass` decorator: 
https://docs.python.org/3.7/library/dataclasses.html

However, it does involve importing `dataclass` from the `dataclasses` library. 

NOTE: Can you check that the default types are suitable, please?

@arporter, let me know what you think of this.

### CODE REVIEW

**Comment by arporter** [comment]
Nice, I didn't know about this. So long as it's standard and works for all our supported versions of Python (and I see that all the tests still pass) then I'm very happy, especially if it lets us document it properly (generating documentation for namedtuples is one of their main weaknesses). Please add a class docstring and see if you can document the various properties so that the reference guide builds correctly and has useful info. in it.

**Comment by sergisiso** [comment]
dataclasses (python>3.7) are fantastic, as well as and typing (python>3.5) which I also see imported in this PR, and it would be great to start using them both!

But given that a release is imminent, maybe we can wait to merge this until after the release is done so we can point to the lastest version that somebody could try using an older python?

One question I have is if we start using typing, e.g. in this PR: `name: str`, do we still need to specify the types in the docstrings, or could we remove them as its duplicated information? Supposedly Sphinx use the type annotations when generating the docs?

**Comment by TeranIvy** [comment]
> > But given that a release is imminent, maybe we can wait to merge this until after the release is done so we can point to the lastest version that somebody could try using an older python?
> 
> Yes, that sounds wise to me so long as @TeranIvy is happy too?

That sounds good, @arporter and @sergisiso! We already have issues with ancient libraries on Crays so I agree with waiting for this after the release. Good work using dataclasses, @oakleybrunt 😄

**Comment by arporter** [comment]
I realised this is in fact the one I should look at next. @oakleybrunt would you mind fixing the conflict and then I'll review.

**Comment by oakleybrunt** [comment]
@arporter it's all yours :)

File: src/psyclone/dynamo0p3.py
Comment (arporter) [inline src/psyclone/dynamo0p3.py:8476]: I don't think this comment is true anymore as a dataclass allows its attributes to be set directly? Therefore, I think you can just delete the sentence beginning "However, ...".

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good. I'll fire off the integration tests. There's just one comment that needs updating I think.

File: src/psyclone/dynamo0p3.py
Comment (oakleybrunt) [reply src/psyclone/dynamo0p3.py]: Indeed, no need for setters with a dataclass.

**Comment by arporter** [comment]
The integration tests threw up the fact that we were inadvertently using Python 3.6 for some of them and that failed here because `dataclass` is only valid in Python >=3.7. Therefore, this PR is now blocked by #2345.

**Comment by arporter** [comment]
As discussed in the meeting this morning, please could you update the implementation of LFRicArgStencil to use a `frozen` dataclass and remove the documentation of the types from the class doc string (since we now have type hints for those).

File: src/psyclone/dynamo0p3.py
Comment (arporter) [reply src/psyclone/dynamo0p3.py]: Well, we could but in this case the code changes required are minimal so I wouldn't bother with implementing setters. (Unless of course we modify any of the members elsewhere too.)

**Comment by oakleybrunt** [comment]
Git sullied my `srcpsyclone/dynamo0p3.py` file when I used `git pull` before pushing. That is why youre seeing three commits with essentailly the same message and a fourth where DynStencil is removed again.

**Review by arporter — APPROVED** [review]
Changes are all good. Integration tests are green.
Thanks for your patience and persistence with this :-)
Will proceed to merge.

**Pull request merged**
